Allow comparison of instructor answers

kitsook opened the pull request:
- Add new column in db to mark if the answer is comparable
- Show a checkbox when an instructor / TA creating a new answer to indicate whether it should be included for student comparisons
- Modify Participation Report for Research Teams to include
instructors/TAs. Dispaly rank and score. Handle multiple answers from
same user.
- Hide user info of answers from Comparison API calls
- Hide answer scores from Comparison API calls
- Add a new API endpoint `/api/courses/:course_id/users/instructionals` to retrieve instructors/TAs of the course

Closes #251, closes #653

kitsook commented:
@andrew-gardener Modified the code as per our discussion.  New column added to indicate if the answer is comparable.

andrew-gardener commented on data/fixtures/test_data.py:
```diff
@@ -465,8 +465,10 @@ def __init__(self):
             self.student_answers.append(answer)
             # add a TA and Instructor answer
             answer = self.create_answer(assignment, self.get_authorized_ta())
```
should add a new answer for both instructor and ta that is comparable (to help testing throughout a variety of situations). will require updating a lot of tests

andrew-gardener commented on compair/static/modules/assignment/assignment-view-partial.html:
```diff
@@ -247,13 +247,29 @@ <h2>All answers
                         <strong>said</strong> on {{answer.created | amDateFormat: 'MMM D @ h:mm a'}}:
                         <div class="manager-actions pull-right" ng-if="canManageAssignment">
                             <em>{{instructors[answer.user_id]}} Response</em>
+                            <em ng-if="answer.comparable">(Comparable)</em>
```
look for `.order_by(desc('instructor_role'), desc('ta_role'))` in `AnswerRootAPI`

kitsook replied:
As discussed, will also change the UI logic to sort instructor answers and student answers together when ordering by answer rank.

Will change the `ng-if` clause to include checking of `answerFilters.orderBy`.
```
                <!-- Instructor Answers Section -->
                <div class="each-answer clearfix" ng-repeat="answer in answers.objects | filter:adminFilter() as adminResults" ng-if="answerFilters.orderBy!='score' && instructors[answer.user_id] && (see_answers || canManageAssignment)">
```

Will also modify the `adminFilter()` function to keep comparable answers from instructors.

andrew-gardener replied:
one thing I missed before is I'm not sure if we want to label the answer as "Comparable" or not. Its something I'd want to check with @lenglund about but since since that will have to wait until next week, I think it would be safer to not include it for now (and add it back later). Maybe make an issue `1.1.1` issue on it

kitsook commented on compair/static/modules/answer/answer-form-partial.html:
```diff
@@ -37,19 +37,30 @@ <h2><i class="fa fa-comments"></i> {{assignment.name}}</h2>
             </select>
         </div>
 
+        <div class="form-group" ng-if="canManageAssignment && answer.draft && comparison_example !== true">
```
will change it to `canManageAssignment && answer.draft && submitAsInstructorOrTA && comparison_example !== true` and explicitly check the draft status too

andrew-gardener replied:
it should show if its a draft or not (just disabled if no longer a draft)

That way instructors know if their answer is being compared or not

kitsook commented on compair/models/comparison.py:
```diff
@@ -109,14 +109,20 @@ def _get_new_comparison_pair(cls, course_id, assignment_id, user_id, pairing_alg
             .filter(and_(
                 UserCourse.course_id == course_id,
-                UserCourse.course_role != CourseRole.student
+                UserCourse.course_role.notin_(
```
Explicitly listing the roles that we want to be included seems safer, e.g. if new roles added in the future

kitsook commented on compair/api/report.py:
```diff
@@ -447,6 +453,102 @@ def peer_feedback_report(course, assignments, group_name):
+def comparison_result_report(title, course, assignments, group_name):
+    report = []
+
+    classlist = UserCourse.query \
```
Safer to state explicitly what we want to retrieve?  e.g. if new roles added in the future, it is easy to see if a report / function not covering the new roles.

andrew-gardener replied:
In this cause I think its making it more complex since you're retrieving them separately and then combining them in two places (and that's all the two separate lists are used for). I don't see why you'd want to run two queries

kitsook commented on compair/api/answer.py:
```diff
@@ -200,6 +202,10 @@ def post(self, course_uuid, assignment_uuid):
+        # instructors / TA can mark the answer as non-comparable
+        if allow(MANAGE, Answer(course_id=course.id)):
+            answer.comparable = False if params.get("comparable") == False else True
+
```
Collectively, with checking on UUID and `else` clause:

```
        # instructor / TA can mark the answer as non-comparable, unless the answer is for a student
        if allow(MANAGE, Answer(course_id=course.id)):
            answer.comparable = False if params.get("comparable") == False and not user_uuid else True
        else:
            answer.comparable = True
```

andrew-gardener replied:
maybe instead check if the `answer.user_id` not equal to current_user.id nad bring it out to the if
```
if allow(MANAGE, Answer(course_id=course.id)) and answer.user_id != current_user.id:
    answer.comparable = params.get("comparable") 
else:
    answer.comparable = True
```

kitsook replied:
But if the instructor submits non-comparable answer for himself, it will be marked incorrectly as comparable

kitsook replied:
should be
```
if allow(MANAGE, Answer(course_id=course.id)) and not user_uuid:
    answer.comparable = params.get("comparable") 
else:
    answer.comparable = True
```

andrew-gardener replied:
sorry meant `answer.user_id == current_user.id`

andrew-gardener replied:
isn't user_uuid sent when instructors create answers for themselves?

kitsook commented on compair/models/assignment.py:
```diff
@@ -278,6 +278,28 @@ def __declare_last__(cls):
+                Answer.draft == False,
+                Answer.practice == False,
+                UserCourse.course_id == cls.course_id,
+                UserCourse.course_role.in_(
```
Checking explicit roles seems safer

kitsook commented:
Referring to @andrew-gardener 's comment on why creating one list for student and one list for instructors/TAs.  The original idea was to filter the student list by group and leave the instructors/TAs intact without filtering.  After discussion, it makes sense to filter the instructors/TAs by group too.  So modified the code to retrieve class users in single query.

https://github.com/ubc/compair/blob/2d62b66f31d7eceb6047294d919e25d81797cc66/compair/api/report.py#L460-L470

kitsook commented:
Added test data for comparable instructor / TA answers.  

https://github.com/ubc/compair/blob/2d62b66f31d7eceb6047294d919e25d81797cc66/data/fixtures/test_data.py#L476-L484

Also modified some test cases because of the added test data.

Commented out this particular test case as the assumption is not always correct.
https://github.com/ubc/compair/blob/2d62b66f31d7eceb6047294d919e25d81797cc66/compair/tests/api/test_comparisons.py#L511-L580

lenglund commented:
Without having looked at this directly yet, my thought is we'll want to treat the compared instructor answer(s) exactly like the student answers. These do not need to be labelled differently (other than adding the instructor label as we already do), and the "Instructor Response" indicator should be replaced with the score/rank.

For the non-compared instructor answer(s) OTOH I think we should keep the "Instructor Response" and add "(Not Compared)" after that. So basically taking Clarence's suggestion but reversing it. Will take a look at the code now.

lenglund commented:
Comments:
* I'm guessing this is just my local dev, but I'm not seeing the instructor answers when I sort by score? Instructor answers are instead appearing only when sorted by date (and without a visible rank/score even when included for comparison)...
* Can we allow for selecting an instructor/TA's name from the filter drop-down?
* On comparisons tab, can we do the same (allow for selecting an instructor/TA's name from the drop-down) and change to "Show instructor's answer" instead of "Show student's answer" for those answers?
*  Can we change the checkbox label to "**Allow this instructor answer to appear in comparison pairings (when the algorithm selects it)**"? Additionally, I think the checkbox should be 1) editable until it's been compared and 2) more clearly disabled when not editable (e.g., "text-muted not-bold" class for label when disabled). 
* Since the default is to include, I'd rather have the instructional text read "Answers submitted as an instructor or TA will be included in comparisons unless this option is unchecked below. All answers will appear to students when the assignment completes."

kitsook commented:
@lenglund 
>I'm guessing this is just my local dev, but I'm not seeing the instructor answers when I sort by score? Instructor answers are instead appearing only when sorted by date (and without a visible rank/score even when included for comparison)...

When sorting by "Answer rank", only answers that are compared (i.e. with score and rank) will be shown.  Possible that in your local dev, the instructor answers are not compared yet?

>Can we allow for selecting an instructor/TA's name from the filter drop-down?

The current `classlist` API has `/students` to retrieve list of student, and `/instructors/labels` to retrieve labels for identifying instructor/TA answers.  But there is no `/instructors` to retrieve list of instructor/TA.  Will add that so that we can filter by instructor/TA.

>On comparisons tab, can we do the same (allow for selecting an instructor/TA's name from the drop-down) and change to "Show instructor's answer" instead of "Show student's answer" for those answers?

Currently on "All Comparisons" tab (only accessible by users that can manage the course), the filter has `All students` and list of individual student.  Will change `All students` to `All` (i.e. to include instructor/TA answers) and add individual instructor/TA

>Can we change the checkbox label to "Allow this instructor answer to appear in comparison pairings (when the algorithm selects it)"?

Will update.

>Additionally, I think the checkbox should be 1) editable until it's been compared and 2) more clearly disabled when not editable (e.g., "text-muted not-bold" class for label when disabled).

For 1), it may complicate the flow as we will need to check for race condition e.g. if an answer is compared when user editing it, user will see error message when trying to save it.  The checkbox will suddenly be disabled and revert to original value.

Suggest to keep the checkbox has the similar logic as the "Submit As" field: that is, only editable when it is a new / draft answer.

For 2), :+1:

>Since the default is to include, I'd rather have the instructional text read "Answers submitted as an instructor or TA will be included in comparisons unless this option is unchecked below. All answers will appear to students when the assignment completes."

Will update :+1:

andrew-gardener commented on compair/tests/api/test_assignment.py:
```diff
@@ -777,7 +777,7 @@ def test_get_all_status(self):
                     self.assertEqual(status['comparisons']['count'], compare_count_result)
                     self.assertEqual(status['comparisons']['left'],
-                        assignment.total_comparisons_required - compare_count_result)
+                        max(0, assignment.total_comparisons_required - compare_count_result))
```
why was this changed? (if compare_count_result is less than zero, there is a problem somewhere)

kitsook replied:
The existing function `_submit_all_possible_comparisons_for_user` compares all possible pairs for a given user without checking how many comparison is required for the assignment.

Will change that logic and fix affected tests.

andrew-gardener replied:
alright, `_submit_all_possible_comparisons_for_user` is working correctly (sorry for not catching that)

andrew-gardener commented on compair/tests/api/test_comparisons.py:
```diff
@@ -314,12 +322,12 @@ def test_get_and_submit_comparison(self, mocked_update_assignment_grades_run, mo
                         new_assignment_grade = AssignmentGrade.get_user_assignment_grade(self.assignment, user)
-                        self.assertGreater(new_course_grade.grade, course_grade)
-                        self.assertGreater(new_assignment_grade.grade, assignment_grade)
+                        self.assertGreaterEqual(new_course_grade.grade, course_grade)
```
this is incorrect, grade should always increase

kitsook replied:
If the student do more comparisons than required (which normally won't happen via GUI), the score won't increase.  The test here is explicitly trying to do all available comparisons and trigger error 400.

Will change the checking to:
- if doing comparisons required by the assignment, grade should increase;
- otherwise the grade should stay the same

andrew-gardener replied:
Thanks, sounds great

andrew-gardener commented on compair/tests/api/test_comparisons.py:
```diff
@@ -501,46 +508,76 @@ def _submit_all_possible_comparisons_for_user(self, user_id):
         }
 
+    """
+    The assumption of the test is not always true, especially when there are
```
The rough assumption of this test should hold (answers that win more should have a higher score). Maybe write the test to capture this better instead of removing it.

andrew-gardener replied:
if there are time pressures, this test can probably be skipped for now though since the algorithm tests also test roughly the same thing

kitsook replied:
Re-enabled the test. Changed the logic slightly to handle cases when answers have the same number of wins.

lenglund commented:
> Possible that in your local dev, the instructor answers are not compared yet?

Yes, but I thought that non-compared answers for students appeared at the bottom of the list with "Not Compared" as text for instructors to see where rank/score would be? I assumed instructor non-compared answers would be the same?

> Will add that so that we can filter by instructor/TA.

I'm envisioning this in the same dropdown; hopefully that works!

> e.g. if an answer is compared when user editing it, user will see error message when trying to save it. The checkbox will suddenly be disabled and revert to original value.

On second thought, how are we handling deleted answers after the comparison period begins? I'd rather leave this checkbox enabled all the time and let instructors add or remove their answer(s) as they want. And if they remove it, we account for it the same way we would a deleted answer from the pool. Could this work?

kitsook commented:
>Yes, but I thought that non-compared answers for students appeared at the bottom of the list with "Not Compared" as text for instructors to see where rank/score would be? I assumed instructor non-compared answers would be the same?

Actually for the existing logic in production, when the "Order by" filter is set as "Answer score", non-compared student answers won't be shown.  When the order filter is set as "Date submitted", all students answers will be shown and those not compared will have status `(not compared)`

The same logic applies to comparable instructor answers.

kitsook commented:
@andrew-gardener test cases added for the new `Comparison Result Report`.  For now the cases only check the number of records and do not verify the content.  @lenglund please see if you have any comment on the report.  Will revise the test cases to add content verification once the layout is confirmed.

Also, added a new API endpoint `/instructionals` in `classlist.py` to retrieve instructors/TAs of the course.

kitsook commented:
@andrew-gardener pls let me know if you want me to merge the branch into master.

kitsook commented:
@lenglund any comments on the new Comparison Result Report?

lenglund commented:
Instead of the column for "User UUID", can we use student number? This is consistent with the other instructor reports. Also, can we change "Answer UUID" to "Answer ID"? Finally, if I remember correctly, we no longer want to show per-criterion scores, so should these columns be removed?

andrew-gardener commented:
@kitsook I'm just realized that the "Participation Report for Research Teams" is very similar to your new report. Why did you add the one exactly? It likely would be better to just add instructors, rankings, answer id, and answer text to existing report 

@lenglund researcher participation report still displays overall score and scores for each criteria

kitsook commented:
@andrew-gardener "Participation Report for Research Teams" report is on a per-user basis.  It lists how many answers and comparisons each user has submitted.  So it can't show the details if instructors submitted multiple answers to see how students compare them differently.

andrew-gardener commented:
@kitsook would it be easy to add multiple rows for instructors?

kitsook commented:
@andrew-gardener It is possible to change the "Participation report for Research Teams" to a one answer per row format.  But this will change the original purpose (measuring user participation?) of the report.

Some columns (e.g. "Evaluations Submitted", "Evaluation Requirements Met", and "Replies Submitted") will remain to be per-user basis.  Some columns will be per-answer basis (e.g. score).

andrew-gardener commented:
@kitsook  "Evaluations Submitted", "Evaluation Requirements Met", "Replies Submitted", and "Student No" should probably be blank for instructors so it should be fine (we don't allow multiple answers from students so it shouldn't be a problem)

kitsook commented:
@andrew-gardener Modified "Participation Report for Research Teams" to include instructors/TAs. Added columns 'Answer', 'Students Ranked', and 'Overall Score'.

Removed "Comparison Result Report".

andrew-gardener commented on compair/api/report.py:
```diff
@@ -134,6 +135,12 @@ def post(self, course_uuid):
             data = peer_feedback_report(course, assignments, group_name)
             titles = [titles1, titles2]
+
+        # elif report_type == "comparison_result":
```
left some old commented out code in

andrew-gardener commented on compair/tests/api/test_comparisons.py:
```diff
@@ -317,12 +325,16 @@ def test_get_and_submit_comparison(self, mocked_update_assignment_grades_run, mo
                     if user.id == self.data.get_authorized_student().id:
                         new_course_grade = CourseGrade.get_user_course_grade(self.course, user)
                         new_assignment_grade = AssignmentGrade.get_user_assignment_grade(self.assignment, user)
-                        self.assertGreater(new_course_grade.grade, course_grade)
```
due to code change you shouldn't need the if anymore (will only go up to max times)
